internal/throttling/throttling.go

```diff
@@ -0,0 +1,249 @@
+package throttling
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-redis/redis/v8"
+
+	"github.com/rudderlabs/rudder-server/services/stats"
+)
+
+/*
+TODOs:
+* benchmark AWS Elasticache Redis
+* replace old limiters
+* guard against concurrency?
+*/
+
+var (
+	//go:embed lua/gcra.lua
+	gcraLua         string
+	gcraRedisScript *redis.Script
+	//go:embed lua/sortedset.lua
+	sortedSetLua    string
+	sortedSetScript *redis.Script
+)
+
+func init() {
+	gcraRedisScript = redis.NewScript(gcraLua)
+	sortedSetScript = redis.NewScript(sortedSetLua)
+}
+
+type redisSpeaker interface {
+	redis.Scripter
+	redisSortedSetRemover
+}
+
+type TokenReturner interface {
+	Return(context.Context) error
+}
+
+type statsCollector interface {
+	NewTaggedStat(name, statType string, tags stats.Tags) stats.Measurement
+}
+
+type Limiter struct {
```

Comment:
Why does `Limiter` need to hold the union of fields for all different algorithms/setups? Wouldn't it be more straightforward if we had (4) different limiter implementations, one for each algorithm and each one containing only the fields it requires?

Reply:
I like the approach I'm proposing because I think it'll be harder to have an implementation go astray from the others and I don't want the consumer side to worry too much about implementation details. With this approach there isn't much complexity anyway, we're talking about a `switch` in `Limit` and a bigger constructor. The constructor returns a pointer to the same `*Limiter` struct with no need to abstract (no interface returned).